### ISSUE TITLE
IINA doesn't open links from other apps

### ISSUE THREAD

**Issue by dillonjohnbrown** [iina/iina#3118]
**System and IINA version:**

- macOS 10.15.6 (19G2021)
- IINA 1.1.0-beta1 Build 121

**Expected behavior:**
IINA opens a URL in a new window (like it does when adding a URL in IINA from File > Open URL...)

**Actual behavior:**
IINA launches (if not open already)

<details>
<summary><b>IINA log:</b></summary>

```
08:54:20.741 [iina][d] App will launch
08:54:20.767 [iina][d] URL event: https://www.youtube.com/watch?v=Irv3KJR6B80
08:54:20.770 [iina][d] App launched
08:54:20.771 [iina][d] Loading key bindings
08:54:20.773 [iina][d] Set key bindings
08:54:20.828 [iina][d] Setting up MediaPlayer integration
08:54:20.832 [player0][d] Set path to /Applications/IINA.app/Contents/MacOS:/usr/bin:/bin:/usr/sbin:/sbin
08:54:20.860 [iina][d] Parsing URL https://www.youtube.com/watch?v=Irv3KJR6B80
08:54:20.868 [player0][d] Track list changed

```
</details>

<br>

<details>
<summary><b>mpv log:</b></summary>

```
[   0.010][v][cplayer] mpv 0.32.0 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
[   0.010][v][cplayer]  built on Thu Aug 27 22:54:26 CST 2020
[   0.010][v][cplayer] ffmpeg library versions:
[   0.010][v][cplayer]    libavutil       56.51.100
[   0.010][v][cplayer]    libavcodec      58.91.100
[   0.010][v][cplayer]    libavformat     58.45.100
[   0.010][v][cplayer]    libswscale      5.7.100
[   0.010][v][cplayer]    libavfilter     7.85.100
[   0.010][v][cplayer]    libswresample   3.7.100
[   0.010][v][cplayer] ffmpeg version: 4.3.1
[   0.010][v][cplayer] 
[   0.010][v][cplayer] Configuration: waf configure --prefix=/usr/local/Cellar/mpv-iina/0.32.0 --enable-javascript --enable-libmpv-shared --enable-lua --enable-libarchive --enable-uchardet --enable-libbluray --disable-swift --disable-debug-build --disable-macos-media-player --confdir=/usr/local/etc/mpv --datadir=/usr/local/Cellar/mpv-iina/0.32.0/share/mpv-iina --mandir=/usr/local/Cellar/mpv-iina/0.32.0/share/man --docdir=/usr/local/Cellar/mpv-iina/0.32.0/share/doc/mpv-iina --lua=51deb
[   0.010][v][cplayer] List of enabled features: 51deb aligned_alloc asm atomics bsd-fstatfs build-date cocoa coreaudio cplayer cplugins drmprime fchmod ffmpeg gl gl-cocoa glob glob-posix gnuc gpl iconv javascript jpeg lcms2 libarchive libass libass-osd libav-any libavcodec libavdevice libavutil libbluray libdl libm libmpv-shared lua macos-10-11-features macos-10-12-2-features macos-10-14-features macos-touchbar optimize osx-thread-name plain-gl posix posix-or-mingw posix-spawn posix-spawn-native pthreads rubberband stdatomic uchardet videotoolbox-gl videotoolbox-hwaccel zlib
[   0.010][v][cplayer] mpv 0.32.0 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
[   0.010][v][cplayer]  built on Thu Aug 27 22:54:26 CST 2020
[   0.010][v][cplayer] ffmpeg library versions:
[   0.010][v][cplayer]    libavutil       56.51.100
[   0.010][v][cplayer]    libavcodec      58.91.100
[   0.010][v][cplayer]    libavformat     58.45.100
[   0.010][v][cplayer]    libswscale      5.7.100
[   0.010][v][cplayer]    libavfilter     7.85.100
[   0.010][v][cplayer]    libswresample   3.7.100
[   0.010][v][cplayer] ffmpeg version: 4.3.1
[   0.010][v][cplayer] 
[   0.010][v][cplayer] Configuration: waf configure --prefix=/usr/local/Cellar/mpv-iina/0.32.0 --enable-javascript --enable-libmpv-shared --enable-lua --enable-libarchive --enable-uchardet --enable-libbluray --disable-swift --disable-debug-build --disable-macos-media-player --confdir=/usr/local/etc/mpv --datadir=/usr/local/Cellar/mpv-iina/0.32.0/share/mpv-iina --mandir=/usr/local/Cellar/mpv-iina/0.32.0/share/man --docdir=/usr/local/Cellar/mpv-iina/0.32.0/share/doc/mpv-iina --lua=51deb
[   0.010][v][cplayer] List of enabled features: 51deb aligned_alloc asm atomics bsd-fstatfs build-date cocoa coreaudio cplayer cplugins drmprime fchmod ffmpeg gl gl-cocoa glob glob-posix gnuc gpl iconv javascript jpeg lcms2 libarchive libass libass-osd libav-any libavcodec libavdevice libavutil libbluray libdl libm libmpv-shared lua macos-10-11-features macos-10-12-2-features macos-10-14-features macos-touchbar optimize osx-thread-name plain-gl posix posix-or-mingw posix-spawn posix-spawn-native pthreads rubberband stdatomic uchardet videotoolbox-gl videotoolbox-hwaccel zlib
[   0.011][v][bdmv/bluray] Opening /Applications/IINA.app/Contents/Resources/config/iina-default-input.conf
[   0.011][v][file] Opening /Applications/IINA.app/Contents/Resources/config/iina-default-input.conf
[   0.011][d][file] resize stream to 131072 bytes
[   0.011][d][file] Stream opened successfully.
[   0.011][v][input] Parsing input config file /Applications/IINA.app/Contents/Resources/config/iina-default-input.conf
[   0.011][v][input] Input config file /Applications/IINA.app/Contents/Resources/config/iina-default-input.conf parsed: 69 binds
[   0.011][d][global] user path: '/Users/dillon/Library/Logs/com.colliderli.iina/2020-09-21-08-54-20_tsf0sE/mpv.log' -> '/Users/dillon/Library/Logs/com.colliderli.iina/2020-09-21-08-54-20_tsf0sE/mpv.log'
[   0.011][d][ytdl_hook] Loading lua script @ytdl_hook.lua...
[   0.011][d][stats] Loading lua script @stats.lua...
[   0.011][d][console] Loading lua script @console.lua...
[   0.012][d][osx] Exiting...
[   0.012][v][cplayer] Waiting for scripts...
[   0.012][v][cplayer] Set property: vo="libmpv" -> 1
[   0.012][d][console] loading mp.defaults
[   0.012][d][stats] loading mp.defaults
[   0.012][d][ytdl_hook] loading mp.defaults
[   0.012][v][cplayer] Set property: keepaspect="no" -> 1
[   0.012][v][cplayer] Set property: gpu-hwdec-interop="auto" -> 1
[   0.013][d][stats] loading @stats.lua
[   0.013][d][console] loading @console.lua
[   0.013][d][ytdl_hook] loading @ytdl_hook.lua
[   0.014][d][stats] reading options for stats 
[   0.014][d][stats] script-opts/stats.conf not found. 
[   0.014][d][stats] lua-settings/stats.conf not found. 
[   0.014][d][ytdl_hook] reading options for ytdl_hook 
[   0.014][d][ytdl_hook] script-opts/ytdl_hook.conf not found. 
[   0.014][d][ytdl_hook] lua-settings/ytdl_hook.conf not found. 
[   0.027][v][ao/coreaudio] selected audio output device: Built-in Output (61)
[   0.028][d][console] reading options for console 
[   0.028][d][console] script-opts/console.conf not found. 
[   0.028][d][console] lua-settings/console.conf not found. 
[   0.029][d][cplayer] Run command: define-section, flags=64, args=["input_stats", "i script-binding stats/display-stats\nI script-binding stats/display-stats-toggle\n", "default"]
[   0.029][d][cplayer] Run command: enable-section, flags=64, args=["input_stats", "allow-hide-cursor+allow-vo-dragging"]
[   0.029][d][cplayer] Run command: define-section, flags=64, args=["input_forced_stats", "", "force"]
[   0.029][d][cplayer] Run command: enable-section, flags=64, args=["input_forced_stats", "allow-hide-cursor+allow-vo-dragging"]
[   0.033][v][cplayer] Done loading scripts.
[   0.034][v][osd/libass] Shaper: FriBidi 1.0.10 (SIMPLE) HarfBuzz-ng 2.7.1 (COMPLEX)
[   0.034][v][osd/libass] Setting up fonts...
[   0.035][v][osd/libass] Using font provider coretext
[   0.035][v][osd/libass] Done.
[   0.036][d][cplayer] Run command: define-section, flags=64, args=["input_console", "", "default"]
[   0.036][d][cplayer] Run command: enable-section, flags=64, args=["input_console", "allow-hide-cursor+allow-vo-dragging"]
[   0.036][d][cplayer] Run command: define-section, flags=64, args=["input_forced_console", "", "force"]
[   0.036][d][cplayer] Run command: enable-section, flags=64, args=["input_forced_console", "allow-hide-cursor+allow-vo-dragging"]

```
</details>

<br>

**Steps to reproduce:**

1. Install [Finicky](https://github.com/johnste/finicky) and set it as default browser
2. Follow Finicky setup instructions and create `.finicky.js` file
3. Inside `.finicky.js` add relevant code so YouTube links open in IINA
4. Click YouTube link in NetNewsWire

<details>
<summary>My .finicky.js contents:</summary>

```
// Generated by Finicky Kickstart 
// Save as ~/.finicky.js

module.exports = {
	defaultBrowser: "Safari",
	rewrite: [
		{
			// Redirect all URLs to use https
			match: ({ url }) => url.protocol === "http",
			url: { protocol: "https" }
		},
		{
			// Remove all marketing/tracking information from URLs
			match: ({ url }) => (url.search.includes("utm_")),
			url({ url }) {
				const search = url.search.split('&').filter(part => !part.startsWith("utm_"));
				return {
					...url,
					search: search.join('&')
				}
			}
		}
	],
	handlers: [
		{
			// Open YouTube links in IINA
			match: /^https:\/\/www\.youtube\.com\/watch..*$/,
			browser: "IINA"
		}
  	]
}
```

</details>

- [ ] MPV does not have this problem.

**How often does this happen?**
Always

**Comment by lhc70000** [comment]
Currently IINA doesn't behave like a browser and it will only handle URLs with the `iina://` scheme. I just added support for it and it will be available soon in 1.1.0 beta 2.

**Comment by dillonjohnbrown** [comment]
You are a saint! Thank you very much for making this change.

**Comment by dillonjohnbrown** [comment]
I just got this feature in an update, and I wanted to thank you again @lhc70000. This change has made IINA and my Mac even better! Thank you for contributing to such an incredible app. The work this group has made is everything good and right about macOS apps.

**Comment by axwnet** [comment]
As of recently, this seems to no longer work.

**Comment by svobs** [comment]
@notthewave can you maybe file a new issue for this? It's likely that it broke in some unexpected way which is unrelated to this issue.

I'm not one of the original developers, so I'm not familiar with this feature, but I have been tinkering with IINA's launch sequence and have found that many things are badly out of date due to changes Apple has made to the interface.

I found [this documentation for the iOS version of the feature](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app) which may or may not work for MacOS. There is [this post on StackOverflow](https://stackoverflow.com/questions/65890597/support-url-schemes-in-macos-application) which makes it seem like IINA must be launched and running before the URL is triggered, because it registers a listener only after it's started. But I did now find [this comment](https://gist.github.com/mralexgray/3915592?permalink_comment_id=2651284#gistcomment-2651284) which suggests that the documentation for the first link will work for MacOS 10.13 and above. But IINA is not (yet) using this API.

Reading through the code, it also looks like IINA is picky about the format of its URLs, and they might need to be escaped. So to open a link for `http://my_website.com/some_video` in a new window in IINA (I used [this tool online](https://www.urlencoder.org/) to do escaping), it looks like the URL sent to IINA might need to be something like `iina://open?url=http%3A%2F%2Fmy_website.com%2Fsome_video&new_window=1`. And it looks like if the syntax is wrong at all the whole thing will be ignored. But I haven't yet tested that. So many things to work on...
